README.md
# JLang JavaScript Plugin

[![GitHub tag](https://img.shields.io/badge/version-1.0.2-blue.svg)](https://github.com/josantonius/js-language-dropdown/tree/1.0.2)
[![Packagist](https://img.shields.io/cocoapods/l/AFNetworking.svg)](https://github.com/josantonius/js-language-dropdown/blob/master/LICENSE)

[Versión en español](README-ES.md)

Language dropdown flags for Materialize and Bootstrap framework.

---

- [Installation](#installation)
- [Requirements](#requirements)
- [Materialize](#materialize)
- [Bootstrap3](#bootstrap3)
- [Bootstrap4](#bootstrap4)
- [Options](#options)
- [Captures](#captures)
- [Usage](#usage)
- [Sponsor](#Sponsor)
- [License](#license)

---

![image](docs/captures/header.png)

---

## Installation

You can download this plugin with [Bower](https://bower.io/).

    bower install jlang --save

The previous command will only install the necessary files, if you prefer to download the entire source code including the demo content you can use:

    git clone https://github.com/josantonius/js-language-dropdown.git

## Requirements

This JavaScript plugin should be used together with the Materialize or Bootstrap framework.

## Materialize

[Live demo](https://josantonius.github.io/js-language-dropdown/materialize.html)

### HTML Structure

```html
<div id="JLang" class="right">
  <div data-lang-code="es-ES" data-src="languages/32/Spain.png">Español</div>
  <div data-lang-code="it-IT" data-src="languages/32/Italy.png">Italian</div>
  <div data-lang-code="en-GB" data-src="languages/32/United-Kingdom.png">English</div>
  <div data-lang-code="de-DE" data-src="languages/32/Germany.png">Deutsche</div>
  <div data-lang-code="fr-FR" data-src="languages/32/France.png">French</div>
</div>
```

### CSS Styles

```css
.language {
  display: -webkit-flex !important;
  display: -ms-flexbox !important;
  display: flex !important;
  -webkit-align-items: center !important;
  -ms-flex-align: center !important;
  align-items: center !important;
}

.language img {
  margin-right: 10px;
}

#lanNavSel {
  position: relative;
  top: -8px;
  left: 8px;
}
```

### JavaScript Inicialization

```js
<script src="jlang.js"></script>
<script>
new JLang();
</script>
```

### JLang Demo

```js
new JLang({
    id: 'languages',
    framework: 'materialize',
    cookieExp: 30,
    cookieLangName: 'lname',
    cookieLangCode: 'lcode',
    abbreviation: true,
    reload: true,
    alignment: 'left',
    hover: true
  }
);
```

## Bootstrap3

[Live demo](https://josantonius.github.io/js-language-dropdown/bootstrap-3.x.html)

### HTML Structure

```html
<div id="JLang">
  <div data-lang-code="es-ES" data-src="languages/32/Spain.png">Español</div>
  <div data-lang-code="it-IT" data-src="languages/32/Italy.png">Italian</div>
  <div data-lang-code="en-GB" data-src="languages/32/United-Kingdom.png">English</div>
  <div data-lang-code="de-DE" data-src="languages/32/Germany.png">Deutsche</div>
  <div data-lang-code="fr-FR" data-src="languages/32/France.png">French</div>
</div>
```

### CSS Styles

```css
.language img {
  margin-right: 10px;
}

#lanNavSel {
  position: relative;
  top: -8px;
  left: 8px;
}
```

### JavaScript Inicialization

```js
<script src="jlang.js"></script>
<script>
new JLang({framework: 'bootstrap3'});
</script>
```

### JLang Demo

```js
new JLang({
    id: 'languages',
    framework: 'bootstrap3',
    cookieExp: 30,
    cookieLangName: 'lname',
    cookieLangCode: 'lcode',
    abbreviation: true,
    reload: true,
    alignment: 'left',
    hover: true
  }
);
```

## Bootstrap4

[Live demo](https://josantonius.github.io/js-language-dropdown/bootstrap-4.x.html)

### HTML Structure

```html
<div id="JLang">
  <div data-lang-code="es-ES" data-src="languages/32/Spain.png">Español</div>
  <div data-lang-code="it-IT" data-src="languages/32/Italy.png">Italian</div>
  <div data-lang-code="en-GB" data-src="languages/32/United-Kingdom.png">English</div>
  <div data-lang-code="de-DE" data-src="languages/32/Germany.png">Deutsche</div>
  <div data-lang-code="fr-FR" data-src="languages/32/France.png">French</div>
</div>
```html

### CSS Styles

```css
.language img {
  margin-right: 10px;
}

#lanNavSel {
  margin-left: 8px;
  margin-right: 8px;
}

#imgNavSel {
  margin-top: -1px;
}

#languages {
  padding: 5px;
}

#languages:hover {
  background-color: rgba(255, 255, 255, 0.15);
}
```

### JavaScript Initialization

```js
<script src="jlang.js"></script>
<script>
new JLang({framework: 'bootstrap4'});
</script>
```

### JLang Demo

```js
new JLang({
    id: 'languages',
    framework: 'bootstrap4',
    cookieExp: 30,
    cookieLangName: 'lname',
    cookieLangCode: 'lcode',
    abbreviation: true,
    reload: true,
    alignment: 'left',
    hover: true
  }
);
```

## Options

| Option Name | Description |
| --- | --- |
| id | Dropdown id name. (Default: 'languages') |
| framework | Framework name. (Default: 'materialize') |
| cookieExp | Cookie expiration time in days. (Default: 30) |
| cookieLangName |  Cookie name for save the selected language name. (Default: 'lname') |
| cookieLangCode | Cookie name for save the selected language code. (Default: 'lcode') |
| abbreviation | Display abbreviation and caret. (Default: true) |
| reload | Reload page when selecting language. (Default: true) |
| alignment | Dropdown id name. (Default: 'languages') |
| hover | Dropdown aligned to the left or right of button. (Default: 'left') |

## Captures

[Materialize](docs/captures/materialize-en.md)

[Bootstrap3](docs/captures/bootstrap3-en.md)

[Bootstrap4](docs/captures/bootstrap4-en.md)

## Sponsor

If this project helps you to reduce your development time,
[you can sponsor me](https://github.com/josantonius#sponsor) to support my open source work :blush:

## License

This repository is licensed under the [MIT License](LICENSE).

Copyright © 2017-2022, [Josantonius](https://github.com/josantonius#contact)


CHANGELOG.md
# CHANGELOG

## 2022-08-20

* The repository was archived.

## 1.0.2 - 2017-04-25

* Changes in documentation.

## 1.0.1 - 2017-04-25

* Changes in documentation.

## 1.0.0 - 2017-04-24

* Added `JLang` class.
* Added `constructor()` method.
* Added `setCookie()` method.
* Added `getCookie()` method.
* Added `checkCookie()` method.
* Added `init()` method.
* Added `getLanguages()` method.
* Added `loadLanguages()` method.
* Added `getActualLanguage()` method.
* Added `setContent()` method.
* Added `appendLanguagesList()` method.
* Added `setNewLanguage()` method.
* Added `dropdownInitialization()` method.


jlang.js
/**
 * Language dropdown flags for Materialize and Bootstrap framework.
 * 
 * @author     Josantonius - hello@josantonius.com
 * @copyright  Copyright (c) 2017
 * @license    https://opensource.org/licenses/MIT - The MIT License (MIT)
 * @link       https://github.com/josantonius/js-language-dropdown
 * @since      1.0.0
 */
class JLang
{

    /**
     * Set parameters and initialize.
     *
     * @since 1.0.0
     */
    constructor(o)
    {

        if (typeof o.id === "undefined") { o.id = "languages"; }
        if (typeof o.framework === "undefined") { o.framework = "materialize"; }
        if (typeof o.cookieExp === "undefined") { o.cookieExp = 30; }
        if (typeof o.cookieLangCode === "undefined") { o.cookieLangCode = "lcode"; }
        if (typeof o.cookieLangName === "undefined") { o.cookieLangName = "lname"; }
        if (typeof o.abbreviation === "undefined") { o.cookieLangName = true; }
        if (typeof o.reload === "undefined") { o.reload = true; }
        if (typeof o.alignment === "undefined") { o.alignment = "left"; }
        if (typeof o.hover === "undefined") { o.hover = true; }

        this.dropdownID = o.id;
        this.framework = o.framework;
        this.cookieExp = o.cookieExp;
        this.cookieLangCode = o.cookieLangCode;
        this.cookieLangName = o.cookieLangName;
        this.abbreviation = o.abbreviation;
        this.reloadPage = o.reload;
        this.alignment = o.alignment;
        this.hover = o.hover;

        this.init();
    }

    /**
     * Set Cookie.
     *
     * @since 1.0.0
     */
    setCookie(cname, cvalue)
    {
        var d = new Date();
        d.setTime(d.getTime() + (this.cookieExp * 24 * 60 * 60 * 1000));
        var expires = "expires=" + d.toUTCString();
        document.cookie = cname + "=" + cvalue + ";" + expires + ";path=/";
    }

    /**
     * Get cookie value.
     *
     * @since 1.0.0
     *
     * @return string
     */
    getCookie(cname)
    {
        var name = cname + "=";
        var ca = document.cookie.split(';');
        for (var i = 0; i < ca.length; i++)
        {
            var c = ca[i];
            while (c.charAt(0) == ' ')
            {
                c = c.substring(1);
            }
            if (c.indexOf(name) == 0)
            {
                return c.substring(name.length, c.length);
            }
        }
        return "";
    }

    /**
     * Check cookie and return value if exists.
     *
     * @since 1.0.0
     *
     * @return string
     */
    checkCookie(cname, cvalue)
    {
        var exists = this.getCookie(cname);
        if (exists === "")
        {
            this.setCookie(cname, cvalue, this.cookieExp);
            return cvalue;
        }
        return exists;
    }

    /**
     * Plugin initializer
     *
     * @since 1.0.0
     */
    init()
    {
        if (this.getLanguages())
        {
            this.loadLanguages();
            if (this.framework === 'materialize')
            {
                this.dropdownInitialization();
            }
        }
    }

    /**
     * Get the languages set in the html code.
     *
     * @since 1.0.0
     */
    getLanguages()
    {
        this.codes = [];
        this.languages = [];
        var jlang = document.getElementById("JLang");
        var elems = jlang.querySelectorAll("[data-lang-code]");
        var count = elems.length;
        if (count)
        {
            for (var i = 0; i < count; i++)
            {
                if (typeof elems[i] === "undefined")
                {
                    break;
                }
                var languageName = elems[i].innerHTML;
                var languageImage = elems[i].getAttribute('data-src');
                var languageCode = elems[i].getAttribute('data-lang-code');
                if (i == 0)
                {
                    this.defaultLang = elems[i].innerHTML;
                    this.defaultCode = languageCode;
                }
                this.languages[languageName] = languageImage;
                this.codes[languageName] = languageCode;
                elems[i].remove;
            }
            return count;
        }
        return false;
    }

    /**
     * Prepare languages to be inserted.
     *
     * @since 1.0.0
     */
    loadLanguages()
    {
        this.actualLanguage = this.getActualLanguage();
        this.setContent();
        this.appendLanguagesList();
        this.setNewLanguage();
    }

    /**
     * Get actual language.
     *
     * @since 1.0.0
     */
    getActualLanguage()
    {
        this.checkCookie(
            this.cookieLangCode,
            this.defaultCode,
            this.cookieExp
        );
        return this.checkCookie(
            this.cookieLangName,
            this.defaultLang,
            this.cookieExp
        );
    }

    /**
     * Load languages listand prepare dropdown according to the framework.
     *
     * @since 1.0.0
     */
    setContent()
    {
        if (this.framework === 'bootstrap3')
        {
            var abbreviation =
                '<span id="lanNavSel"></span>' +
                '<span class="caret"></span>';
            var content =
                '<ul class="nav navbar-nav navbar-right">' +
                '<li class="dropdown">';
            var data = 'class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false"';
            var openTag = '<li>';
            var closeTag = '</li>';
            var startContent = '<ul id="languages-list" class="dropdown-menu">';
            var endContent = '</ul></li></ul>';
        } else if (this.framework === 'bootstrap4')
        {
            var abbreviation =
                '<span id="lanNavSel"></span>' +
                '<span class="caret"></span>';
            var content =
                '<ul class="navbar-nav">' +
                '<li class="nav-item dropdown">';
            var data = 'class="nav-link dropdown-toggle" href="#!" id="languages-list" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"';
            var openTag = '';
            var closeTag = '';
            var startContent = '<div id="languages-list" class="dropdown-menu" aria-labelledby="dropdown-list">';
            var endContent = '</div></li></ul>';
        } else if (this.framework === 'materialize')
        {
            var abbreviation =
                '<span id="lanNavSel" class="white-text"></span>' +
                '<i class="material-icons white-text right">arrow_drop_down</i>';
            var content = '';
            var data = 'class="dropdown-button btn-flat" data-activates="languages-list"';
            var openTag = '<li>';
            var closeTag = '</li>';
            var startContent = '<ul id="languages-list" class="dropdown-content">';
            var endContent = '</ul>';
        }
        abbreviation = (this.abbreviation) ? abbreviation : '';
        this.content =
            content +
            '<a href="#" id="' + this.dropdownID + '"' + data + '>' +
            '<img id="imgNavSel" src="" alt="" class="flag-icon">' +
            abbreviation +
            '</a>' +
            startContent;
        var key;
        var languages = this.languages;
        var languagesList = '';
        for (key in languages)
        {
            if (languages.hasOwnProperty(key))
            {
                var image = languages[key];
                languagesList +=
                    openTag +
                    '<a id="nav' + key + '" href="#" class="language dropdown-item">' +
                    '<img id="imgNav' + key + '" src="' + image + '" alt="' + key + '" class="flag-icon">' +
                    '<span id="lanNav' + key + '"> ' + key + '</span>' +
                    '</a>' +
                    closeTag;
            }
        }
        this.languagesList = languagesList + endContent;
    }

    /**
     * Insert languages.
     *
     * @since 1.0.0
     */
    appendLanguagesList()
    {
        var JLang = document.getElementById('JLang');
        JLang.innerHTML = this.content;
        var languagesList = document.getElementById('languages-list');
        languagesList.innerHTML = this.languagesList;
        var language = this.actualLanguage.toUpperCase();
        if (this.abbreviation)
        {
            var lanNavSel = document.getElementById('lanNavSel');
            lanNavSel.innerHTML = ' ' + language.substring(0, 3);
        }
        var imgNavSel = document.getElementById('imgNavSel');
        var image = this.languages[this.actualLanguage];
        imgNavSel.setAttribute('src', image);
    }

    /**
     * Save the new selected language.
     *
     * @since 1.0.0
     */
    setNewLanguage()
    {
        var elems = document.getElementsByClassName('language');
        var count = elems.length;
        var obj = this;
        for (var i = count; i--;)
        {
            elems[i].addEventListener('click', function ()
            {
                var currentId = this.getAttribute('id');
                if (~currentId.indexOf('nav'))
                {
                    var type = 'Nav';
                    var language = currentId.replace('nav', '');
                }
                var img = document.getElementById('img' + type + 'Sel');
                var imgLink = this.childNodes[0];
                var imgLink = imgLink.getAttribute('src');
                img.setAttribute('src', imgLink);
                var upperLang = language.toUpperCase();

                if (obj.abbreviation)
                {
                    var lanNavSel = document.getElementById('lanNavSel');
                    lanNavSel.innerHTML = ' ' + upperLang.substring(0, 3);
                    lanNavSel.setAttribute('alt', language);
                }
                obj.setCookie(obj.cookieLangName, language, obj.cookieExp);
                obj.setCookie(obj.cookieLangCode, obj.codes[language], obj.cookieExp);
                if (obj.reloadPage)
                {
                    location.reload();
                }
            }, false);
        }
    }

    /**
     * Inializer for dropdown in Materialize framework.
     *
     * @since 1.0.0
     */
    dropdownInitialization()
    {
        $('#' + this.dropdownID).dropdown({
            inDuration: 300,
            outDuration: 225,
            constrainWidth: false,
            hover: this.hover,
            gutter: 0,
            belowOrigin: false,
            alignment: this.alignment,
            stopPropagation: false
        }
        );
    }
}